File: srcp/day-17.py
from srcp.utils.input import get_data
import re

data = get_data(2024, 17)

lines: list[str] = data.splitlines()

register_1: int = int(re.compile(r'Register A: (\d+)').findall(lines[0])[0])
register_2: int = int(re.compile(r'Register B: (\d+)').findall(lines[1])[0])
register_3: int = int(re.compile(r'Register C: (\d+)').findall(lines[2])[0])
# Programme toujours : [2, 4, 1, 3, 7, 5, 1, 5, 0, 3, 4, 2, 5, 5, 3, 0]
program = list(map(int, re.compile(r'\d+').findall(lines[4])))
target_output = ",".join(map(str, program))

opcode: dict = {0: 'adv', 1: 'bxl', 2: 'bst', 3: 'jnz', 4: 'bxc', 5: 'out', 6: 'bdv', 7: 'cdv'}


def get_combo(op: int, a: int, b: int, c: int) -> int:
    if op == 4:
        return a
    elif op == 5:
        return b
    elif op == 6:
        return c
    return op


def dv(register: int, op: int, a: int, b: int, c: int) -> int:
    return register >> get_combo(op, a, b, c)


def get_output(prog: list[int], a: int = 0, b: int = 0, c: int = 0) -> list[int]:
    pointer: int = 0
    outs: list[int] = []

    while pointer < len(prog):
        instruction: str = opcode[prog[pointer]]
        operand = prog[pointer + 1]
        combo = get_combo(operand, a, b, c)
        match instruction:
            case 'adv':
                a = dv(a, operand, a, b, c)
            case 'bxl':
                b ^= operand
            case 'bst':
                b = combo & 7
            case 'jnz':
                if a != 0:
                    pointer = operand
                    continue
            case 'bxc':
                b ^= c
            case 'out':
                outs.append(combo & 7)
            case 'bdv':
                b = dv(a, operand, a, b, c)
            case 'cdv':
                c = dv(a, operand, a, b, c)
        # print(f"Pointer: {pointer}, A: {a}, B: {b}, C: {c}, Output: {outs}")
        pointer += 2
    return outs


# partie 1
outputs = get_output(program, register_1, register_2, register_3)
print(",".join(map(str, outputs)))

# Partie 2

candidates: list[int] = [0]
print(program)
for i in range(1, len(program) + 1):
    print("------")
    print(f"Pour {program[-i:]}")
    out = []
    for c in candidates:
        print(f"Candidat : {c}")
        for offset in range(2 ** 3):
            #  2**3 -> 8 dernier bits
            a = (2 ** 3) * c + offset
            outs = get_output(program, a, register_2, register_3)
            if outs == program[-i:]:
                print(f"A valid pour {program[-i:]} == {outs} -> {a}")
                out.append(a)
    candidates = out

print(min(candidates))

File: src/2024/day17.ts
export function solve_day_17(input: string): number {
    const lines: string[] = input.split('\n');

    const register_a: number = parseInt(lines[0].match(/Register A: (\d+)/)![1]);
    const register_b: number = parseInt(lines[1].match(/Register B: (\d+)/)![1]);
    const register_c: number = parseInt(lines[2].match(/Register C: (\d+)/)![1]);
    const program: number[] = lines[4].split(' ')[1].split(',').map(x => parseInt(x));

    const get_combo = (op: number, a: number, b: number, c: number): number => {
        if (op == 4) {
            return a;
        } else if (op == 5) {
            return b;
        } else if (op == 6) {
            return c;
        }
        return op;
    }

    const dv = (register: number, op: number, a: number, b: number, c: number): number => {
        return register >> get_combo(op, a, b, c);
    }

    const get_outputs = (prog: number[], a: number = 0, b: number = 0, c: number = 0): bigint[] => {
        let pointer: number = 0;
        let outputs: bigint[] = [];

        while (pointer < prog.length) {
            const instruction: number = prog[pointer];
            const operand: number = prog[pointer + 1];
            const tmp_combo: number = get_combo(operand, a, b, c);
            const tmp_dv: number = dv(a, operand, a, b, c);
            switch (instruction) {
                case 0:
                    a = tmp_dv;
                    break;
                case 1:
                    b ^= operand;
                    break;
                case 2:
                    b = tmp_combo & 7;
                    break;
                case 3:
                    if (a !== 0) {
                        pointer = operand;
                        continue;
                    }
                    break;
                case 4:
                    b ^= c;
                    break;
                case 5:
                    const val: bigint = BigInt(tmp_combo) % BigInt(8);
                    outputs.push(val);
                    break;
                case 6:
                    c = tmp_dv;
                    break;
                case 7:
                    c = tmp_dv;
                    break;
            }
            //console.log(`Pointer ${pointer}, A ${a}, B ${b}, C ${c}, Output ${outputs.join(',')}`);
            pointer += 2;
        }
        return outputs;
    }

    // Partie 1
    const part_1: bigint[] = get_outputs(program, register_a, register_b, register_c);
    console.log(`Partie 1 : ${part_1.join(',')}`);

    // Partie 2
    let candidates: number[] = [0];

    for (let i = 1; i <= program.length; i++) {
        let out: number[] = [];
        //console.log(`Pour ${program.slice(-i)}`);

        for (let c of candidates) {
            //console.log(`Candidat : ${c}`);
            for (let offset = 0; offset < 2 ** 3; offset++) {
                const a: number = (2 ** 3) * c + offset;
                const tmp_outputs: bigint[] = get_outputs(program, a, register_b, register_c);
                if (tmp_outputs.join(",") == program.slice(-i).join(",")) {
                    //console.log(`A valid pour ${program.slice(-i)} == ${tmp_outputs} -> ${a}`);
                    out.push(a);
                }
            }
        }
        candidates = out;
    }
    console.log(`Partie 2 : ${candidates.join(',')}`);
    return 0;
}